<?xml version="1.0" ?>
<launch>
  
  <!--Set to false if running online on Lolo-->
  <param name ="/use_sim_time" value="true"/>

	<!-- If mode = sim: Simulated AUV with MBES and control in loaded bathymetric map 
	If mode = gt: reproduce Hugin survey and construct map of real bathymetry --> 
  <arg name="mode" default="gt"/>
	<arg name="namespace" default="lolo"/>
	<arg name="debug" default="0"/>
	<arg name="dataset" default="overnight_2020"/>
	<arg name="app" default="rbpf"/>
  <arg name="path" default="$(find uw_tests)/datasets/$(arg dataset)/"/>

  <!-- MBES sensor -->
	<arg name="n_beams_mbes" default="50"/>	
	<arg name="mbes_open_angle" value="1.050" /> 

  <!-- Coordinates of odom frame wrt the map frame. By default, the AUV start the mission here -->
	<arg name="x" default="0.0"/>
	<arg name="y" default="0.0"/>
	<arg name="z" default="0.0"/>
	<arg name="roll" default="0.0"/>
	<arg name="pitch" default="0.0"/>
	<arg name="yaw" default="0.0"/>

  <!-- RBPF params  -->
  <arg name="particle_count"                    default="3"/> <!--50 -->
  <!-- [x, y, z, roll, pitch, yaw] -->
  <!-- <arg name="init_covariance" default="[0., 0., 0.0, 0.0, 0.0, 0.0]"/> -->
  <arg name="init_covariance" default="[10., 5., 0.0, 0.0, 0.0, 0.0]"/>
  <arg name="motion_covariance" default="[0.0, 0.0, 0.0, 0.0, 0.0, 0.0000]"/>
  <arg name="resampling_noise_covariance" default="[2.0, 2.0, 0.0, 0.0, 0.0, 0.0]"/>
  <!-- <arg name="motion_covariance" default="[0.0, 0.0, 0.0, 0.0, 0.0, 0.001]"/> -->
  <!-- <arg name="resampling_noise_covariance" default="[10, 10, 0.0, 0.0, 0.0, 0.0]"/>  -->
  <arg name="measurement_std"  default="3"/>
  <arg name="survey_finished_top" default="/gt/survey_finished"/>
  <arg name="gp_plot_server" default="/gp/plot_server"/>
  <arg name="sample_plot_server" default="/gp/sample_server"/>
  <arg name="results_path"   default="$(find uw_tests)/$(arg app)/$(arg dataset)/" /> <!-- Where to save the results  -->
  <arg name="rbpf_period"   default="0.3" />


  <arg name="map_frame"                         default="utm"/>
  <!-- <arg name="map_frame"                         default="map"/> -->
  <arg name="odom_frame"                        default="lolo/odom"/>
  <arg name="odometry_topic"                    default="/lolo/dr/odom"/>
  <arg name="mbes_pings_topic"                  default="/lolo/mbes/odom/bathy_points"/>
  <!-- <arg name="mbes_pings_topic"                  default="/lolo/mbes/bathy_points"/> -->
  <arg name="average_pose_topic"                default="/$(arg mode)/avg_pose"/>
  <arg name="particle_poses_topic"              default="/$(arg mode)/particle_poses"/>
  <arg name="particle_sim_mbes_topic"           default="/$(arg mode)/sim_mbes"/>
  <arg name="average_mbes_topic"                default="/$(arg mode)/avg_mbes"/>

  <!-- <node type="static_transform_publisher" pkg="tf" name="tf_map_to_loloodom" args="0 0 0 0 0 0 $(arg map_frame) lolo/odom 10"/> -->
  <node type="static_transform_publisher" pkg="tf" name="tf_utm_to_map" args="0 0 0 0 0 0 $(arg map_frame) map 10"/>
  <node type="static_transform_publisher" pkg="tf" name="tf_map_to_loloodom" args="0 0 0 0 0 0 map $(arg odom_frame) 10"/>

  <include file="$(find lolo_description)/launch/lolo_description.launch" />

  <group ns="$(arg app)">
    <group ns="$(arg namespace)">
      <node type="rbpf_slam.py" pkg="rbpf_slam" name="rbpf_slam" output="screen">
        <param name="particle_count"          type= "int"     value="$(arg particle_count)" />
        <param name="init_covariance"                         value="$(arg init_covariance)" />
        <param name="resampling_noise_covariance"             value="$(arg resampling_noise_covariance)" />
        <param name="measurement_std"                         value="$(arg measurement_std)" />
        <param name="motion_covariance"                       value="$(arg motion_covariance)" />
        <param name="odometry_topic"                          value="$(arg odometry_topic)" />
        <param name="mbes_pings_topic"                        value="$(arg mbes_pings_topic)" />
        <param name="map_frame"                               value="$(arg map_frame)" />
        <param name="odom_frame"                              value="$(arg odom_frame)" />
        <param name="mbes_link"                               value="$(arg namespace)/mbes_link" />
        <param name="base_link"                               value="$(arg namespace)/base_link" />
        <param name="synch_topic"                             value="/$(arg app)/synch" />
        <param name="result_path"                             value="$(arg results_path)" />
        <param name="mbes_open_angle"                         value="$(arg mbes_open_angle)" />
        <param name="average_pose_topic"                      value="$(arg average_pose_topic)" />
        <param name="average_mbes_topic"                      value="$(arg average_mbes_topic)" />
        <param name="particle_poses_topic"                    value="$(arg particle_poses_topic)" />
        <param name="particle_sim_mbes_topic"                 value="$(arg particle_sim_mbes_topic)" />
        <param name="num_beams_sim"                           value="$(arg n_beams_mbes)" />
        <param name="n_beams_mbes"                            value="$(arg n_beams_mbes)" />       
        <param name="mesh_path" value="$(arg path)/mesh.npz" />       
        <param name="survey_finished_top" value="/gt/survey_finished" />       
        <param name="sound_velocity_prof" value="$(arg path)/svp.cereal" />       
        <param name="pf_stats_top" value="/stats/pf_data" />  
        <param name="plot_gp_server" value="$(arg gp_plot_server)"/>     
        <param name="sample_gp_server" value="$(arg sample_plot_server)"/>  
        <param name="rbpf_period" value="$(arg rbpf_period)"/>  
      </node>
    </group>
  </group>

    <!-- Each GP in now trained in a ROS node. This launches a GP ROS node per particle  -->
    <node type="particles_launcher.py" pkg="rbpf_slam" name="particle_launcher" output="screen">
      <param name="num_particles" value="$(arg particle_count)"/>     
      <param name="particle_launch_file" value="$(find rbpf_slam)/launch/particle.launch"/>     
    </node>

    <!-- <node type="visual_tools.py" pkg="rbpf_slam" name="visual_tools" output="screen">
      <param name="pf_stats_top" value="/stats/pf_data" />       
      <param name="map_frame" value="$(arg map_frame)" />       
      <param name="odom_frame" value="$(arg odom_frame)" />       
      <param name="survey_finished_top" value="$(arg survey_finished_top)" />       
      <param name="background_img_path" value="/home/nvidia/catkin_ws/src/UWExploration/utils/uw_tests/datasets/overnight_2020/default_real_mean_depth.png" />       
      <param name="result_path" value="$(arg results_path)" />
      <param name="mbes_pings_topic" value="/$(arg mode)/mbes_pings" />
      <param name="particle_sim_mbes_topic" value="$(arg particle_sim_mbes_topic)" />
   </node> -->
    
</launch>

